krun.py

```diff
@@ -203,13 +213,50 @@ def get_overall_time_estimate_formatter(self):
     def add_eta_info(self, key, exec_time):
         self.eta_estimates[key].append(exec_time)
 
-    def run(self):
+    def build_schedule(self, config, current_result_json):
+        one_exec_scheduled = False
+        eta_avail_job = None
+        for exec_n in xrange(config["N_EXECUTIONS"]):
+            for vm_name, vm_info in config["VMS"].items():
+                for bmark, param in config["BENCHMARKS"].items():
+                    for variant in vm_info["variants"]:
+                        job = ExecutionJob(self, config, vm_name, vm_info, bmark, variant, param)
+                        if not util.should_skip(config, job.key):
+                            if one_exec_scheduled and not eta_avail_job:
+                                # first job of second executions eta becomes known.
+                                eta_avail_job = job
+                                self.set_eta_avail()
+                            self.add_job(job)
+                        else:
+                            if not one_exec_scheduled:
+                                debug("DEBUG: %s is in skip list. Not scheduling." %
+                                      job.key)
+            one_exec_scheduled = True
+        # Resume mode: if previous results are available, remove the
+        # jobs from the schedule which have already been executed, and
+        # add the results to this object, ready to be saved to a Json file.
+        if self.resume and current_result_json is not None:
+            for key in current_result_json['data']:
+                if len(current_result_json['data'][key]) > 0:
+                    try:
+                        self.remove_job_by_key(key)
+                        debug("DEBUG: %s has already been run. Not scheduling." %
+                               key)
+                    except JobMissingError as excn:
+                        tup = (excn.key, self.config_file, self.out_file)
+                        msg = ("Failed to resume benchmarking session\n." +
+                               "The execution %s appears in results " +
+                               "file: %s, but not in config file: %s." % tup)
+                        util.fatal(msg)
+                    self.results[key] = current_result_json['data'][key]
+
+    def run(self, dry_run=False):
         """Benchmark execution starts here"""
 
-        log_and_mail(self.mailer, info,
-                     "Benchmarking started",
-                     "Benchmarking started.\nLogging to %s" % self.log_path,
-                     bypass_limiter=True)
+        util.log_and_mail(self.mailer, info,
```

Comment:
One final thing I remembered last night. We are going to get an email every time krun starts? We should only get an email saying benchmarking started if not in resume mode?